ccl/sqlccl: TestBackupRestoreStatementResult failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/8ef84ea309e3fd54e18f2496794c73484e732a19

Parameters:
```
TAGS=deadlock
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=346148&tab=buildLog

```

I170907 08:03:59.021825 32 server/config.go:520  [n?] 1 storage engine initialized
I170907 08:03:59.021945 32 server/config.go:522  [n?] RocksDB cache size: 128 MiB
I170907 08:03:59.022017 32 server/config.go:522  [n?] store 0: in-memory, size 100 MiB
I170907 08:03:59.031562 32 server/node.go:344  [n?] **** cluster 51bb45d9-d949-4b83-8449-7c2764aa71ac has been created
I170907 08:03:59.031642 32 server/server.go:833  [n?] **** add additional nodes by specifying --join=127.0.0.1:41711
I170907 08:03:59.033809 32 storage/store.go:1196  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170907 08:03:59.033962 32 server/node.go:461  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=3.2 KiB), ranges=1, leases=0, writes=0.00, bytesPerReplica={p10=3266.00 p25=3266.00 p50=3266.00 p75=3266.00 p90=3266.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00}
I170907 08:03:59.034077 32 server/node.go:326  [n1] node ID 1 initialized
I170907 08:03:59.034243 32 gossip/gossip.go:311  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:41711" > attrs:<> locality:<> ServerVersion:<major_val:1 minor_val:0 patch:0 unstable:3 > 
I170907 08:03:59.034564 32 storage/stores.go:303  [n1] read 0 node addresses from persistent storage
I170907 08:03:59.034737 32 server/node.go:606  [n1] connecting to gossip network to verify cluster ID...
I170907 08:03:59.036639 32 server/node.go:631  [n1] node connected via gossip and verified as part of cluster "51bb45d9-d949-4b83-8449-7c2764aa71ac"
I170907 08:03:59.036729 32 server/node.go:403  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I170907 08:03:59.036871 32 sql/executor.go:402  [n1] creating distSQLPlanner with address {tcp 127.0.0.1:41711}
I170907 08:03:59.040007 32 server/server.go:925  [n1] starting https server at 127.0.0.1:41836
I170907 08:03:59.040077 32 server/server.go:926  [n1] starting grpc/postgres server at 127.0.0.1:41711
I170907 08:03:59.040135 32 server/server.go:927  [n1] advertising CockroachDB node at 127.0.0.1:41711
W170907 08:03:59.040215 32 sql/jobs/registry.go:156  [n1] unable to get node liveness: node not in the liveness table
I170907 08:03:59.051564 1390 storage/replica_command.go:2710  [split,n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
POTENTIAL DEADLOCK: Inconsistent locking. saw this ordering in one goroutine:
happened before
../../storage/store.go:1104 storage.(*Store).Start { s.mu.Lock() } <<<<<
../../server/node.go:451 server.(*Node).startStores { if err := s.Start(ctx, stopper); err != nil { }
../../server/node.go:383 server.(*Node).start { if err := n.startStores(ctx, stores, n.stopper); err != nil { }
../../server/server.go:871 server.(*Server).Start { ) }
../../server/testserver.go:323 server.(*TestServer).Start { if err := ts.Server.Start(context.Background()); err != nil { }
../../testutils/serverutils/test_server_shim.go:198 serverutils.StartServerRaw { if err := server.Start(args); err != nil { }
../../testutils/serverutils/test_server_shim.go:168 serverutils.StartServer { server, err := StartServerRaw(params) }
../../testutils/testcluster/testcluster.go:248 testcluster.(*TestCluster).doAddServer { s, conn, _ := serverutils.StartServer(t, serverArgs) }
../../testutils/testcluster/testcluster.go:156 testcluster.StartTestCluster { if err := tc.doAddServer(t, serverArgs); err != nil { }
backup_test.go:83 sqlccl_test.backupRestoreTestSetupWithParams { tc = testcluster.StartTestCluster(t, clusterSize, params) }
backup_test.go:141 sqlccl_test.backupRestoreTestSetup { return backupRestoreTestSetupWithParams(t, clusterSize, numAccounts, init, base.TestClusterArgs{}) }
backup_test.go:199 sqlccl_test.TestBackupRestoreStatementResult { _, dir, _, sqlDB, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, initNone) }

happened after
../../storage/timedmutex.go:92 storage.(*timedMutex).Lock { tm.mu.Lock() } <<<<<
../../storage/replica.go:627 storage.(*Replica).init { r.raftMu.Lock() }
../../storage/replica.go:621 storage.NewReplica { return r, r.init(desc, store.Clock(), replicaID) }
../../storage/store.go:1117 storage.(*Store).Start.func1 { rep, err := NewReplica(&desc, s, 0) }
../../storage/store.go:1034 storage.IterateRangeDescriptors.func1 { return fn(desc) }
../../storage/engine/mvcc.go:1708 engine.MVCCIterate { done, err := f(roachpb.KeyValue{Key: metaKey.Key, Value: *value}) }
../../storage/store.go:1038 storage.IterateRangeDescriptors { false /* !reverse */, kvToDesc) }
../../storage/store.go:1147 storage.(*Store).Start { }) }
../../server/node.go:451 server.(*Node).startStores { if err := s.Start(ctx, stopper); err != nil { }
../../server/node.go:383 server.(*Node).start { if err := n.startStores(ctx, stores, n.stopper); err != nil { }
../../server/server.go:871 server.(*Server).Start { ) }
../../server/testserver.go:323 server.(*TestServer).Start { if err := ts.Server.Start(context.Background()); err != nil { }
../../testutils/serverutils/test_server_shim.go:198 serverutils.StartServerRaw { if err := server.Start(args); err != nil { }
../../testutils/serverutils/test_server_shim.go:168 serverutils.StartServer { server, err := StartServerRaw(params) }
../../testutils/testcluster/testcluster.go:248 testcluster.(*TestCluster).doAddServer { s, conn, _ := serverutils.StartServer(t, serverArgs) }
../../testutils/testcluster/testcluster.go:156 testcluster.StartTestCluster { if err := tc.doAddServer(t, serverArgs); err != nil { }
backup_test.go:83 sqlccl_test.backupRestoreTestSetupWithParams { tc = testcluster.StartTestCluster(t, clusterSize, params) }
backup_test.go:141 sqlccl_test.backupRestoreTestSetup { return backupRestoreTestSetupWithParams(t, clusterSize, numAccounts, init, base.TestClusterArgs{}) }
backup_test.go:199 sqlccl_test.TestBackupRestoreStatementResult { _, dir, _, sqlDB, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, initNone) }

in another goroutine: happened before
../../storage/timedmutex.go:92 storage.(*timedMutex).Lock { tm.mu.Lock() } <<<<<
../../storage/replica.go:3142 storage.(*Replica).handleRaftReady { r.raftMu.Lock() }
../../storage/store.go:3548 storage.(*Store).processReady { stats, err := r.handleRaftReady(IncomingSnapshot{}) }
../../storage/scheduler.go:216 storage.(*raftScheduler).worker { s.processor.processReady(ctx, id) }
../../storage/scheduler.go:167 storage.(*raftScheduler).Start.func2 { s.worker(ctx) }
../../util/stop/stopper.go:198 stop.(*Stopper).RunWorker.func1 { f(ctx) }

happend after
../../storage/store.go:3841 storage.(*Store).tryGetOrCreateReplica { s.mu.Lock() } <<<<<
../../storage/store.go:3760 storage.(*Store).getOrCreateReplica { ) }
../../storage/replica.go:4422 storage.(*Replica).acquireSplitLock { rightRng, created, err := r.store.getOrCreateReplica(ctx, split.RightDesc.RangeID, 0, nil) }
../../storage/replica.go:4412 storage.(*Replica).maybeAcquireSplitMergeLock { return r.acquireSplitLock(ctx, &split.SplitTrigger) }
../../storage/replica.go:4280 storage.(*Replica).processRaftCommand { if splitMergeUnlock := r.maybeAcquireSplitMergeLock(ctx, raftCmd); splitMergeUnlock != nil { }
../../storage/replica.go:3434 storage.(*Replica).handleRaftReadyRaftMuLocked { if changedRepl := r.processRaftCommand(ctx, commandID, e.Term, e.Index, command); changedRepl { }
../../storage/replica.go:3144 storage.(*Replica).handleRaftReady { return r.handleRaftReadyRaftMuLocked(inSnap) }
../../storage/store.go:3548 storage.(*Store).processReady { stats, err := r.handleRaftReady(IncomingSnapshot{}) }
../../storage/scheduler.go:216 storage.(*raftScheduler).worker { s.processor.processReady(ctx, id) }
../../storage/scheduler.go:167 storage.(*raftScheduler).Start.func2 { s.worker(ctx) }
../../util/stop/stopper.go:198 stop.(*Stopper).RunWorker.func1 { f(ctx) }


Other goroutines holding locks:
goroutine 1429 lock 0xc4201c0378
../../server/status/recorder.go:361 status.(*MetricsRecorder).WriteStatusSummary { mr.writeSummaryMu.Lock() } <<<<<
../../server/node.go:754 server.(*Node).writeSummaries.func1 { err = n.recorder.WriteStatusSummary(ctx, n.storeCfg.DB) }
../../util/stop/stopper.go:231 stop.(*Stopper).RunTask { f(ctx) }
../../server/node.go:755 server.(*Node).writeSummaries { }); runErr != nil { }
../../server/node.go:731 server.(*Node).startWriteSummaries.func1 { if err := n.writeSummaries(ctx); err != nil { }
../../util/stop/stopper.go:198 stop.(*Stopper).RunWorker.func1 { f(ctx) }

goroutine 1131 lock 0xc420241578
../../storage/timedmutex.go:92 storage.(*timedMutex).Lock { tm.mu.Lock() } <<<<<
../../storage/replica.go:3142 storage.(*Replica).handleRaftReady { r.raftMu.Lock() }
../../storage/store.go:3548 storage.(*Store).processReady { stats, err := r.handleRaftReady(IncomingSnapshot{}) }
../../storage/scheduler.go:216 storage.(*raftScheduler).worker { s.processor.processReady(ctx, id) }
../../storage/scheduler.go:167 storage.(*raftScheduler).Start.func2 { s.worker(ctx) }
../../util/stop/stopper.go:198 stop.(*Stopper).RunWorker.func1 { f(ctx) }

goroutine 1131 lock 0xc42033b7a8
../../storage/timedmutex.go:92 storage.(*timedMutex).Lock { tm.mu.Lock() } <<<<<
../../storage/store.go:3836 storage.(*Store).tryGetOrCreateReplica { repl.raftMu.Lock() }
../../storage/store.go:3760 storage.(*Store).getOrCreateReplica { ) }
../../storage/replica.go:4422 storage.(*Replica).acquireSplitLock { rightRng, created, err := r.store.getOrCreateReplica(ctx, split.RightDesc.RangeID, 0, nil) }
../../storage/replica.go:4412 storage.(*Replica).maybeAcquireSplitMergeLock { return r.acquireSplitLock(ctx, &split.SplitTrigger) }
../../storage/replica.go:4280 storage.(*Replica).processRaftCommand { if splitMergeUnlock := r.maybeAcquireSplitMergeLock(ctx, raftCmd); splitMergeUnlock != nil { }
../../storage/replica.go:3434 storage.(*Replica).handleRaftReadyRaftMuLocked { if changedRepl := r.processRaftCommand(ctx, commandID, e.Term, e.Index, command); changedRepl { }
../../storage/replica.go:3144 storage.(*Replica).handleRaftReady { return r.handleRaftReadyRaftMuLocked(inSnap) }
../../storage/store.go:3548 storage.(*Store).processReady { stats, err := r.handleRaftReady(IncomingSnapshot{}) }
../../storage/scheduler.go:216 storage.(*raftScheduler).worker { s.processor.processReady(ctx, id) }
../../storage/scheduler.go:167 storage.(*raftScheduler).Start.func2 { s.worker(ctx) }
../../util/stop/stopper.go:198 stop.(*Stopper).RunWorker.func1 { f(ctx) }

goroutine 1390 lock 0xc420240a90
../../storage/queue.go:559 storage.(*baseQueue).processReplica { bq.processMu.Lock() } <<<<<
../../storage/queue.go:530 storage.(*baseQueue).processLoop.func1.2 { if err := bq.processReplica(annotatedCtx, repl, clock); err != nil { }
../../util/stop/stopper.go:231 stop.(*Stopper).RunTask { f(ctx) }
../../storage/queue.go:539 storage.(*baseQueue).processLoop.func1 { }) != nil { }
../../util/stop/stopper.go:198 stop.(*Stopper).RunWorker.func1 { f(ctx) }




ERROR: exit status 2

context canceled
Makefile:260: recipe for target 'stress' failed
make: *** [stress] Error 1
```

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -1101,7 +1101,6 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
 	// (consistent=false). Uncommitted intents which have been abandoned
 	// due to a split crashing halfway will simply be resolved on the
 	// next split attempt. They can otherwise be ignored.
-	s.mu.Lock()
 
 	// TODO(peter): While we have to iterate to find the replica descriptors
 	// serially, we can perform the migrations and replica creation
@@ -1118,9 +1117,17 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
 			if err != nil {
 				return false, err
 			}
-			if err = s.addReplicaInternalLocked(rep); err != nil {
+
+			// We can't lock s.mu across NewReplica due to the lock ordering
+			// constraint (*Replica).raftMu < (*Store).mu. See the comment on
+			// (Store).mu.
+			s.mu.Lock()
+			err = s.addReplicaInternalLocked(rep)
+			s.mu.Unlock()
+			if err != nil {
 				return false, err
 			}
+
 			// Add this range and its stats to our counter.
 			s.metrics.ReplicaCount.Inc(1)
 			s.metrics.addMVCCStats(rep.GetMVCCStats())
@@ -1145,7 +1152,6 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
 			// and initialize those groups.
 			return false, nil
 		})
-	s.mu.Unlock()
 	if err != nil {
 		return err
 	}

